File: src/main/java/dev/robothanzo/werewolf/service/impl/PlayerServiceImpl.java
package dev.robothanzo.werewolf.service.impl;

import dev.robothanzo.werewolf.database.documents.Session;
import dev.robothanzo.werewolf.security.SessionRepository;
import dev.robothanzo.werewolf.service.DiscordService;
import dev.robothanzo.werewolf.service.GameSessionService;
import dev.robothanzo.werewolf.service.PlayerService;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.Optional;

@Slf4j
@Service
@RequiredArgsConstructor
public class PlayerServiceImpl implements PlayerService {

    private final SessionRepository sessionRepository;
    private final DiscordService discordService;
    private final GameSessionService gameSessionService;

    @Override
    public List<Map<String, Object>> getPlayersJSON(long guildId) {
        Optional<Session> sessionOpt = sessionRepository.findByGuildId(guildId);
        if (sessionOpt.isEmpty()) {
            throw new RuntimeException("Session not found");
        }
        return gameSessionService.playersToJSON(sessionOpt.get());
    }

    @Override
    public void setPlayerCount(long guildId, int count, java.util.function.Consumer<String> onProgress,
                               java.util.function.Consumer<Integer> onPercent) throws Exception {
        var session = sessionRepository.findByGuildId(guildId)
                .orElseThrow(() -> new RuntimeException("Session not found"));

        var jda = discordService.getJDA();
        var guild = jda.getGuildById(guildId);
        if (guild == null) {
            throw new Exception("Guild not found");
        }

        if (onPercent != null)
            onPercent.accept(0);
        if (onProgress != null)
            onProgress.accept("開始同步 Discord 狀態...");

        Map<String, Session.Player> players = session.getPlayers();
        List<dev.robothanzo.werewolf.utils.DiscordActionRunner.ActionTask> deleteTasks = new ArrayList<>();
        List<dev.robothanzo.werewolf.utils.DiscordActionRunner.ActionTask> createRoleTasks = new ArrayList<>();
        List<dev.robothanzo.werewolf.utils.DiscordActionRunner.ActionTask> createChannelTasks = new ArrayList<>();

        // Phase 1: Identify deletions
        var existingPlayerIds = new java.util.LinkedList<>(players.keySet());
        for (String idStr : existingPlayerIds) {
            int pid = Integer.parseInt(idStr);
            if (pid > count) {
                var player = players.remove(idStr);
                var role = guild.getRoleById(player.getRoleId());
                if (role != null) {
                    deleteTasks.add(new dev.robothanzo.werewolf.utils.DiscordActionRunner.ActionTask(role.delete(),
                            "刪除身分組: " + role.getName()));
                }
                var channel = guild.getTextChannelById(player.getChannelId());
                if (channel != null) {
                    deleteTasks.add(new dev.robothanzo.werewolf.utils.DiscordActionRunner.ActionTask(channel.delete(),
                            "刪除頻道: " + channel.getName()));
                }
            }
        }

        // Run Deletions (0% -> 30%)
        if (!deleteTasks.isEmpty()) {
            dev.robothanzo.werewolf.utils.DiscordActionRunner.runActions(deleteTasks, onProgress, onPercent, 0, 30, 60);
        }

        // Phase 2: Create Roles
        var spectatorRole = guild.getRoleById(session.getSpectatorRoleId());
        Map<Integer, net.dv8tion.jda.api.entities.Role> newRolesMap = new java.util.concurrent.ConcurrentHashMap<>();

        for (int i = players.size() + 1; i <= count; i++) {
            int playerId = i;
            String name = "玩家" + dev.robothanzo.werewolf.database.documents.Session.Player.ID_FORMAT.format(i);
            var task = new dev.robothanzo.werewolf.utils.DiscordActionRunner.ActionTask(
                    guild.createRole()
                            .setColor(dev.robothanzo.werewolf.utils.MsgUtils.getRandomColor())
                            .setHoisted(true)
                            .setName(name),
                    "創建身分組: " + name);
            task.onSuccess = (obj) -> newRolesMap.put(playerId, (net.dv8tion.jda.api.entities.Role) obj);
            createRoleTasks.add(task);
        }

        if (!createRoleTasks.isEmpty()) {
            dev.robothanzo.werewolf.utils.DiscordActionRunner.runActions(createRoleTasks, onProgress, onPercent, 30, 60,
                    60);
        }

        // Phase 3: Create Channels
        for (var entry : newRolesMap.entrySet()) {
            int playerId = entry.getKey();
            var role = entry.getValue();
            String name = "玩家" + dev.robothanzo.werewolf.database.documents.Session.Player.ID_FORMAT.format(playerId);

            var task = new dev.robothanzo.werewolf.utils.DiscordActionRunner.ActionTask(
                    guild.createTextChannel(name)
                            .addPermissionOverride(spectatorRole != null ? spectatorRole : guild.getPublicRole(),
                                    net.dv8tion.jda.api.Permission.VIEW_CHANNEL.getRawValue(),
                                    net.dv8tion.jda.api.Permission.MESSAGE_SEND.getRawValue())
                            .addPermissionOverride(role,
                                    List.of(net.dv8tion.jda.api.Permission.VIEW_CHANNEL,
                                            net.dv8tion.jda.api.Permission.MESSAGE_SEND),
                                    List.of())
                            .addPermissionOverride(guild.getPublicRole(),
                                    List.of(),
                                    List.of(net.dv8tion.jda.api.Permission.VIEW_CHANNEL,
                                            net.dv8tion.jda.api.Permission.MESSAGE_SEND,
                                            net.dv8tion.jda.api.Permission.USE_APPLICATION_COMMANDS)),
                    "創建頻道: " + name);

            task.onSuccess = (obj) -> {
                var channel = (net.dv8tion.jda.api.entities.channel.concrete.TextChannel) obj;
                players.put(String.valueOf(playerId),
                        dev.robothanzo.werewolf.database.documents.Session.Player.builder()
                                .id(playerId)
                                .roleId(role.getIdLong())
                                .channelId(channel.getIdLong())
                                .build());
            };
            createChannelTasks.add(task);
        }

        if (!createChannelTasks.isEmpty()) {
            dev.robothanzo.werewolf.utils.DiscordActionRunner.runActions(createChannelTasks, onProgress, onPercent, 60,
                    95, 120);
        }

        session.setPlayers(players);
        sessionRepository.save(session);
        if (onPercent != null)
            onPercent.accept(100);
        if (onProgress != null)
            onProgress.accept("同步完成！");
        gameSessionService.broadcastUpdate(guildId);
    }

    @Override
    public void updatePlayerRoles(long guildId, String playerId, List<String> roles) {
        try {
            var session = sessionRepository.findByGuildId(guildId)
                    .orElseThrow(() -> new RuntimeException("Session not found"));
            var player = session.getPlayers().get(playerId);
            if (player == null)
                throw new Exception("Player not found");

            List<String> finalRoles = new ArrayList<>(roles);
            boolean isDuplicated = roles.contains("複製人");
            player.setDuplicated(isDuplicated);
            if (isDuplicated && finalRoles.size() == 2) {
                if (finalRoles.get(0).equals("複製人"))
                    finalRoles.set(0, finalRoles.get(1));
                else if (finalRoles.get(1).equals("複製人"))
                    finalRoles.set(1, finalRoles.get(0));
            }

            player.setIdiot(finalRoles.contains("白癡"));
            boolean isJinBaoBao = session.isDoubleIdentities() && finalRoles.size() == 2 &&
                    finalRoles.get(0).equals("平民") && finalRoles.get(1).equals("平民");
            player.setJinBaoBao(isJinBaoBao);
            player.setRoles(finalRoles);

            sessionRepository.save(session);

            var jda = discordService.getJDA();
            if (jda != null && player.getChannelId() != 0) {
                var guild = jda.getGuildById(guildId);
                if (guild != null) {
                    var channel = guild.getTextChannelById(player.getChannelId());
                    if (channel != null) {
                        channel.sendMessage("法官已將你的身份更改為: " + String.join(", ", roles)).queue();
                    }
                }
            }

            gameSessionService.broadcastUpdate(guildId);
        } catch (Exception e) {
            log.error("Failed to update player roles: {}", e.getMessage(), e);
            throw new RuntimeException("Failed to update player roles", e);
        }
    }

    @Override
    public void switchRoleOrder(long guildId, String playerId) {
        try {
            var session = sessionRepository.findByGuildId(guildId)
                    .orElseThrow(() -> new RuntimeException("Session not found"));
            var player = session.getPlayers().get(playerId);
            if (player == null)
                throw new Exception("Player not found");

            if (player.isRolePositionLocked())
                throw new Exception("你的身分順序已被鎖定");
            if (player.getRoles() == null || player.getRoles().size() < 2)
                throw new Exception("Not enough roles to switch");

            java.util.Collections.swap(player.getRoles(), 0, 1);
            sessionRepository.save(session);

            var jda = discordService.getJDA();
            if (jda != null && player.getChannelId() != 0) {
                var guild = jda.getGuildById(guildId);
                if (guild != null) {
                    var channel = guild.getTextChannelById(player.getChannelId());
                    if (channel != null) {
                        channel.sendMessage("你已交換了角色順序，現在主要角色為: " + player.getRoles().get(0)).queue();
                    }
                }
            }
            gameSessionService.broadcastUpdate(guildId);
        } catch (Exception e) {
            log.error("Switch role order failed: {}", e.getMessage(), e);
            throw new RuntimeException("Failed to switch role order", e);
        }
    }

    @Override
    public void setRolePositionLock(long guildId, String playerId, boolean locked) {
        try {
            var session = sessionRepository.findByGuildId(guildId)
                    .orElseThrow(() -> new RuntimeException("Session not found"));
            var player = session.getPlayers().get(playerId);
            if (player == null)
                throw new Exception("Player not found");

            player.setRolePositionLocked(locked);
            sessionRepository.save(session);
            gameSessionService.broadcastUpdate(guildId);
        } catch (Exception e) {
            log.error("Failed to set role position lock: {}", e.getMessage(), e);
            throw new RuntimeException("Failed to set role position lock", e);
        }
    }
}

File: src/main/kotlin/dev/robothanzo/werewolf/service/impl/PlayerServiceImpl.kt
package dev.robothanzo.werewolf.service.impl

import dev.robothanzo.werewolf.database.SessionRepository
import dev.robothanzo.werewolf.database.documents.Player
import dev.robothanzo.werewolf.database.documents.Session
import dev.robothanzo.werewolf.service.DiscordService
import dev.robothanzo.werewolf.service.GameSessionService
import dev.robothanzo.werewolf.service.PlayerService
import dev.robothanzo.werewolf.utils.ActionTask
import dev.robothanzo.werewolf.utils.MsgUtils
import dev.robothanzo.werewolf.utils.runActions
import net.dv8tion.jda.api.Permission
import net.dv8tion.jda.api.entities.Role
import net.dv8tion.jda.api.entities.channel.concrete.TextChannel
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Service
import java.util.concurrent.ConcurrentHashMap

@Service
class PlayerServiceImpl(
    private val sessionRepository: SessionRepository,
    private val discordService: DiscordService,
    private val gameSessionService: GameSessionService
) : PlayerService {
    private val log = LoggerFactory.getLogger(PlayerServiceImpl::class.java)

    override fun getPlayersJSON(session: Session): List<Map<String, Any>> {
        return gameSessionService.playersToJSON(session)
    }

    @Throws(Exception::class)
    override fun setPlayerCount(
        session: Session,
        count: Int,
        onProgress: (String) -> Unit,
        onPercent: (Int) -> Unit
    ) {
        val guildId = session.guildId
        val jda = discordService.jda
        val guild = jda.getGuildById(guildId) ?: throw Exception("Guild not found")

        onPercent(0)
        onProgress("開始同步 Discord 狀態...")

        val players = session.players
        val deleteTasks = mutableListOf<ActionTask>()
        val createRoleTasks = mutableListOf<ActionTask>()
        val createChannelTasks = mutableListOf<ActionTask>()

        // Phase 1: Identify deletions
        val existingPlayerIds = players.keys.toMutableList()
        for (idStr in existingPlayerIds) {
            val pid = idStr.toInt()
            if (pid > count) {
                val player = players.remove(idStr)
                if (player != null) {
                    val role = guild.getRoleById(player.roleId)
                    if (role != null) {
                        deleteTasks.add(
                            ActionTask(
                                role.delete(),
                                "刪除身分組: ${role.name}"
                            )
                        )
                    }
                    val channel = guild.getTextChannelById(player.channelId)
                    if (channel != null) {
                        deleteTasks.add(
                            ActionTask(
                                channel.delete(),
                                "刪除頻道: ${channel.name}"
                            )
                        )
                    }
                }
            }
        }
        // Run Deletions (0% -> 30%)
        if (deleteTasks.isNotEmpty()) {
            deleteTasks.runActions(onProgress, onPercent, 0, 30, 60)
        }

        // Phase 2: Create Roles
        val spectatorRole = guild.getRoleById(session.spectatorRoleId)
        val newRolesMap: MutableMap<Int, Role> = ConcurrentHashMap()

        for (i in players.size + 1..count) {
            val name = "玩家" + Player.ID_FORMAT.format((i).toLong())
            val task = ActionTask(
                guild.createRole()
                    .setColor(MsgUtils.randomColor)
                    .setHoisted(true)
                    .setName(name),
                "創建身分組: $name",
                { obj: Any? -> newRolesMap[i] = obj as Role }
            )
            createRoleTasks.add(task)
        }

        if (createRoleTasks.isNotEmpty()) {
            createRoleTasks.runActions(onProgress, onPercent, 30, 60, 60)
        }

        // Phase 3: Create Channels
        for ((playerId, role) in newRolesMap) {
            val name = "玩家" + Player.ID_FORMAT.format((playerId).toLong())

            val task = ActionTask(
                guild.createTextChannel(name)
                    .addPermissionOverride(
                        spectatorRole ?: guild.publicRole,
                        Permission.VIEW_CHANNEL.rawValue,
                        Permission.MESSAGE_SEND.rawValue
                    )
                    .addPermissionOverride(
                        role,
                        listOf(
                            Permission.VIEW_CHANNEL,
                            Permission.MESSAGE_SEND
                        ),
                        listOf()
                    )
                    .addPermissionOverride(
                        guild.publicRole,
                        listOf(),
                        listOf(
                            Permission.VIEW_CHANNEL,
                            Permission.MESSAGE_SEND,
                            Permission.USE_APPLICATION_COMMANDS
                        )
                    ),
                "創建頻道: $name",
                { obj: Any? ->
                    val channel = obj as TextChannel
                    players[playerId.toString()] = Player(
                        id = playerId,
                        roleId = role.idLong,
                        channelId = channel.idLong
                    )
                }
            )
            createChannelTasks.add(task)
        }

        if (createChannelTasks.isNotEmpty()) {
            createChannelTasks.runActions(onProgress, onPercent, 60, 95, 120)
        }

        session.players = players
        sessionRepository.save(session)
        onPercent(100)
        onProgress("同步完成！")
    }

    override fun updatePlayerRoles(player: Player, roles: List<String>) {
        try {
            val session = player.session ?: throw Exception("Player session not found")
            
            val finalRoles = roles.toMutableList()
            val isDuplicated = roles.contains("複製人")
            player.duplicated = isDuplicated
            if (isDuplicated && finalRoles.size == 2) {
                if (finalRoles[0] == "複製人")
                    finalRoles[0] = finalRoles[1]
                else if (finalRoles[1] == "複製人")
                    finalRoles[1] = finalRoles[0]
            }

            player.idiot = finalRoles.contains("白癡")
            val isJinBaoBao =
                session.doubleIdentities && finalRoles.size == 2 && finalRoles[0] == "平民" && finalRoles[1] == "平民"
            player.jinBaoBao = isJinBaoBao
            player.roles = finalRoles

            sessionRepository.save(session)

            val jda = discordService.jda
            if (jda != null && player.channelId != 0L) {
                val guild = jda.getGuildById(session.guildId)
                if (guild != null) {
                    player.send("法官已將你的身份更改為: ${roles.joinToString(", ")}")
                }
            }
        } catch (e: Exception) {
            log.error("Failed to update player roles: {}", e.message, e)
            throw RuntimeException("Failed to update player roles", e)
        }
    }

    override fun switchRoleOrder(player: Player) {
        try {
            val session = player.session ?: throw Exception("Player session not found")

            if (player.rolePositionLocked)
                throw Exception("你的身分順序已被鎖定")

            val roles = player.roles
            if (roles == null || roles.size < 2)
                throw Exception("Not enough roles to switch")

            roles.let {
                val first = it[0]
                it[0] = it[1]
                it[1] = first
            }
            sessionRepository.save(session)
            player.send("你已交換了角色順序，現在主要角色為: ${roles[0]}")
        } catch (e: Exception) {
            log.error("Switch role order failed: {}", e.message, e)
            throw RuntimeException("Failed to switch role order", e)
        }
    }

    override fun setRolePositionLock(player: Player, locked: Boolean) {
        try {
            val session = player.session ?: throw Exception("Player session not found")

            player.rolePositionLocked = locked
            sessionRepository.save(session)
        } catch (e: Exception) {
            log.error("Failed to set role position lock: {}", e.message, e)
            throw RuntimeException("Failed to set role position lock", e)
        }
    }
}
